include/multipass/utils.h

```diff
@@ -127,10 +129,69 @@ std::string qenum_to_string(RegisteredQtEnum val);
 // other helpers
 QString get_multipass_storage();
 QString make_uuid(const std::optional<std::string>& seed = std::nullopt);
+
 template <typename OnTimeoutCallable, typename TryAction, typename... Args>
 void try_action_for(OnTimeoutCallable&& on_timeout, std::chrono::milliseconds timeout, TryAction&& try_action,
                     Args&&... args);
 
+template <typename T>
+bool is_default_constructed(const T& input_type)
+{
+    return input_type == T{};
+}
+
+// simplified parallel transform, it takes a std container and a unary operation and
+// returns a std::vector<OutputValueType> where the OutputValueType is the unary operation return type
+// There are two options of the return types, one is the one below and the other one is auto. Eventually, I went with
+// the std::invoke_result_t based one because it makes the function signature more expressive despite the fact that it
+// makes std::invoke_result_t<std::decay_t<UnaryOperation>, InputValueType> code duplicate.
+template <typename Container, typename UnaryOperation>
+std::vector<std::invoke_result_t<std::decay_t<UnaryOperation>, typename Container::value_type>>
```

Comment:
It looks like this type, as you mentioned, is duplicated and is hard to read.  Any way to instead add the `using`s before this and use them here and in the function?

Reply:
Good doubt you have raised, I am afraid there is a proper way to do this. The challenge here is that the `using` can only called in the function body in the scope of this function and we need that alias on the declaration of the function which is before the function body. 

However, there is a workaround. If we define the alias out of this scope by using template alias. We will have the result code like this:
```
template <typename Container, typename UnaryOperation>
using OutputValueType = std::invoke_result_t<std::decay_t<UnaryOperation>, typename Container::value_type>;

template <typename Container, typename UnaryOperation>
std::vector<OutputValueType<Container, UnaryOperation>>
parallel_transform(const Container& input_container, UnaryOperation&& unary_op)
{
    using OutputValueType = OutputValueType<Container, UnaryOperation>;
    const auto num_elements = input_container.size();

    // Pre-allocate space for futures
    std::vector<std::future<OutputValueType>> futures(num_elements);
    ...
}
```
The downside here is that the duplication of the `template <typename Container, typename UnaryOperation>` scope and the scope of the alias `OutputValueType` is no longer within the `parallel_transform` function. Plus, the user of the alias has to call like `OutputValueType<Container, UnaryOperation>` with the specified template parameter. 

As to which one to choose, the original one is the same as `std::async`'s function's signature and it only has one template scope, on the other side, the new one does eliminate the duplication. I do not have a strong opinion on this. If you prefer the new one, then I will change it.

Reply:
After considering what you've said, I think we should leave it as it is now.  Thanks!